support-frontend/app/actions/AsyncAuthenticatedBuilder.scala

```diff
@@ -0,0 +1,41 @@
+package actions
+
+import com.typesafe.scalalogging.LazyLogging
+import play.api.mvc.Security.AuthenticatedRequest
+import play.api.mvc._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+// scalastyle:off
+// This is based on AuthenticationBuilder in https://github.com/playframework/playframework/blob/1ca9d0af237ac45a011671b2e036d726cd05e4e7/core/play/src/main/scala/play/api/mvc/Security.scala
+// The main difference is that this action builder enables authenticating a user asynchronously.
+// i.e. RequestHeader => Future[Option[U]] instead of RequestHeader => Option[U]
+// In the context of Guardian applications, asynchronous authentication equals making a call to identity API,
+// to authenticate the user data.
+// TODO: considering porting this to identity-play-auth.
+// scalastyle:on
+class AsyncAuthenticatedBuilder[U](
+    userinfo: RequestHeader => Future[Option[U]],
+    defaultParser: BodyParser[AnyContent],
+    onUnauthorized: RequestHeader => Result
+)(implicit val executionContext: ExecutionContext)
+  extends ActionBuilder[({ type R[A] = AuthenticatedRequest[A, U] })#R, AnyContent] with LazyLogging { // scalastyle:ignore
+
+  lazy val parser = defaultParser
+
+  def invokeBlock[A](request: Request[A], block: AuthenticatedRequest[A, U] => Future[Result]): Future[Result] =
+    authenticate(request, block)
+
+  /**
+    * Authenticate the given block.
+    */
+  def authenticate[A](request: Request[A], block: AuthenticatedRequest[A, U] => Future[Result]): Future[Result] = {
+    userinfo(request).flatMap {
+      case Some(user) => block(new AuthenticatedRequest(user, request))
+      case None => Future.successful(onUnauthorized(request))
```

Comment:
Not recommending a change, but here's an interesting friday topic:
the interesting thing is that this copy of the class is needed because the monad to use (option) is not a type parameter, so the logic isn't reusable.  Not that I'm suggestion the tagless final approach!
However an interesting point is that in some way map/getOrElse (or fold could be used) is more abstract than pattern match, as pattern match only works on ADTs wheras the high level methods work on OptionT as well https://typelevel.org/cats/datatypes/optiont.html